### DISCUSSION TITLE
How to update logging config properly when tenant is switched

### DISCUSSION THREAD

**Discussion by dirkam** [spatie/laravel-multitenancy#388] [Q&A] [answered]
I use a custom `SwitchTenantTask` to update some of the config parameters when tenant is switched to update the filesystem settings, etc.
I also update the logging to have a separate log file for each tenant:
```php
public function makeCurrent(Tenant $tenant): void
    {
        config()->set('logging.channels.daily.path', storage_path("app/tenant-{$tenant->id}/logs/laravel.log"));
    }
```
This works, logs are going to the dedicated path.
However, if I use the `Log` facade prior to this, then all logs are using the default path.
```php
public function makeCurrent(Tenant $tenant): void
    {
        Log::debug('test');
        config()->set('logging.channels.daily.path', storage_path("app/tenant-{$tenant->id}/logs/laravel.log"));
    }
```

I also tried to add the provided `ClearFacadeInstancesTask` but that doesn't seem to help with this.

Can someone point me in the right direction understanding this behavior? I worry that this might happen with other things as well that are called prior switch tenant tasks.

**Comment by masterix21** [comment, marked as answer]
What about a task to create an on-demand channel?

https://laravel.com/docs/9.x/logging#on-demand-channels

**Reply by dirkam** [reply]
@masterix21 Thanks for your response.
That'd certainly work, but I'd prefer to use the default channel so it doesn't have to be changed everywhere and it would ensure that the logs always go to the correct path (it's easy to forget to use a specific channel).

More importantly, I'd like to understand why this is happening. I'm afraid that this behavior might affect other things as well, I just happened to run into it with the `Log` facade.

**Reply by dirkam** [reply]
What seems to be working is that I created a new `tenant` channel in `logging.php`. In the task, I update the config values of this channel. And I also set the default config from daily to tenant. This way the logs prior to tenant switch go to the daily channel and the tenant-related logs go to the tenant channel.
Still not sure if this works in all cases.

**Reply by masterix21** [reply]
why shouldn't?

**Reply by dirkam** [reply]
Well, it should. But since I don't fully get why the first approach was not working, I'm not sure if this one has any surprises. Thank you for having a look.

**Reply by masterix21** [reply]
Surprises are behind any corner, but if your application is well-tested, you are a step over.
